### PR TITLE
Feature/cleanup

### PR DESCRIPTION
**Pull request by prudhomm**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes?
- [ ] Have you successfully run the Feel++ testsuite with your changes locally?

-----
* implements #1084 (instantiate Mesh<>) to hopefully reduce compilation cost by reducing template instantiation
* a bit of cleanup (eg remove multi_index headers...)
* Localization class is no more a subclass of Mesh but a template class of its own, cleaned up the code a bit where Localization is used

### CODE REVIEW

**Comment by prudhomm** [comment]
here is the testsuite
```
94% tests passed, 24 tests failed out of 370

Label Time Summary:
testalg              =  34.84 sec (22 tests)
testcore             =  52.78 sec (42 tests)
testcrb              = 227.97 sec (17 tests)
testdiscr            = 1008.74 sec (109 tests)
testfilters          = 366.95 sec (15 tests)
testfit              =   9.18 sec (4 tests)
testintegration      = 325.86 sec (14 tests)
testinterpolation    = 104.24 sec (25 tests)
testleaks            =  26.62 sec (16 tests)
testls               =   4.64 sec (3 tests)
testmaterial         =   2.03 sec (2 tests)
testmath             =   2.44 sec (2 tests)
testmesh             =  55.91 sec (26 tests)
testmodels           =   5.04 sec (2 tests)
testopt              =   5.11 sec (4 tests)
testpde              =  63.23 sec (4 tests)
testpoly             =  52.49 sec (18 tests)
testts               =  24.29 sec (7 tests)
testvf               = 406.97 sec (38 tests)

Total Test time (real) = 863.50 sec

The following tests FAILED:
	 39 - feelpp_test_hdf5-np-6 (Failed)
	 41 - feelpp_test_serialization-np-6 (Failed)
	 42 - feelpp_test_serialization-np-1 (Failed)
	111 - feelpp_test_importergmsh-np-6 (Timeout)
	112 - feelpp_test_importergmsh-np-1 (Failed)
	123 - feelpp_test_med-np-1 (Failed)
	157 - feelpp_test_integration-np-6 (Timeout)
	158 - feelpp_test_integration-np-1 (Failed)
	186 - feelpp_test_space_0d-np-6 (Not Run)
	187 - feelpp_test_space_0d-np-1 (Not Run)
	254 - feelpp_test_nlSolveComposite-np-6 (Failed)
	255 - feelpp_test_nlSolveComposite-np-1 (Failed)
	262 - feelpp_test_markerhdf5_2D-np-6 (Timeout)
	264 - feelpp_test_markerhdf5_3D-np-6 (Timeout)
	279 - feelpp_test_eim_grepl-np-1 (Timeout)
	289 - feelpp_test_db-np-6 (Failed)
	299 - feelpp_test_bdf3-np-1 (Not Run)
	308 - feelpp_test_laplacianv-np-6 (Not Run)
	309 - feelpp_test_laplacianv-np-1 (Not Run)
	310 - feelpp_test_laplacian-np-6 (Not Run)
	311 - feelpp_test_laplacian-np-1 (Not Run)
	312 - feelpp_test_laplaciant-np-6 (Not Run)
	313 - feelpp_test_laplaciant-np-1 (Not Run)
	324 - feelpp_test_convolve-np-6 (Timeout)
Errors while running CTest
```

**Comment by prudhomm** [comment]
@vincentchabannes @Trophime  I see a lot of problems with HDF5 1.8 which I don't see with HDF5 1.10. @vincentchabannes  are you testing feelpp testsuite against HDF5 1.8?

recall that hdf5 1.8 is required to have proper med support but then if it means that we can no longer properly handle our own hdf5 files we have a problem!

**Comment by Trophime** [comment]
The question, here is more concerning `hdf5` compatibily.
 I guess you can view files generated by `1.8 hdf5` version with `1.10` but not the inverse.

There is a likely workaround which consists in discarding, for the moment, med support.
This means we need to have a way to convert med into `msh` format... This would require an additional
docker/singularity images. This will more things to manage ...
Remind that this situation is directly connected with a compatibility issue between `med` file format and `hdf5 1.10`. This should be solved by the end of this year with `med-fichier` 4.xx versions.

To sum up we have 3 choices:

* [ ] stick to `hdf5 1.8`

* [ ] move to `hdf5 1.10`, discard `med` support

* [ ] same as above but we provide external tools to convert `med` into `msh`.
 
For this last two solutions, we may end up with trouble if we use recent *Debian* / *Ubuntu* os because of singularity issue.  Remind that *Debian Stretch* / *Ubuntu Xenial* singularity containers are working fine on FTII contrary to newer os versions because of GLIBC incompatibility

**Comment by vincentchabannes** [comment]
testsuite on atlas : 

100% tests passed, 0 tests failed out of 369

Label Time Summary:
testalg              =  72.63 sec (22 tests)
testcore             = 122.04 sec (42 tests)
testcrb              = 204.66 sec (15 tests)
testdiscr            = 666.80 sec (111 tests)
testfilters          =  86.71 sec (14 tests)
testfit              =  13.31 sec (4 tests)
testintegration      =  52.02 sec (14 tests)
testinterpolation    = 127.53 sec (25 tests)
testleaks            =  52.95 sec (16 tests)
testls               =  10.66 sec (3 tests)
testmaterial         =   4.99 sec (2 tests)
testmath             =   5.37 sec (2 tests)
testmesh             =  88.19 sec (26 tests)
testmodels           =   6.66 sec (2 tests)
testopt              =  11.92 sec (4 tests)
testpde              =  19.62 sec (4 tests)
testpoly             =  64.01 sec (18 tests)
testts               =  22.90 sec (7 tests)
testvf               = 138.65 sec (38 tests)

Total Test time (real) = 1778.19 sec

The following tests did not run:
	185 - feelpp_test_space_0d-np-6 (Skipped)
	186 - feelpp_test_space_0d-np-1 (Skipped)
	298 - feelpp_test_bdf3-np-1 (Skipped)
	307 - feelpp_test_laplacianv-np-6 (Skipped)
	308 - feelpp_test_laplacianv-np-1 (Skipped)
	309 - feelpp_test_laplacian-np-6 (Skipped)
	310 - feelpp_test_laplacian-np-1 (Skipped)
	311 - feelpp_test_laplaciant-np-6 (Skipped)
	312 - feelpp_test_laplaciant-np-1 (Skipped)
	323 - feelpp_test_convolve-np-6 (Skipped)
